tests/test_solvers.py

```diff
@@ -0,0 +1,295 @@
+            }))
+
+    @contextlib.contextmanager
+    def env(self, key, value):
+        old = os.environ.get(key)
+        os.environ[key] = value
+        try:
+            yield
+        except:
+            if old is not None:
+                os.environ[key] = old
+
+    @contextlib.contextmanager
+    def solver(self, *, add=(), remove=(), packages=()):
+        """Create a new solver.
+
+        Roughly equivalent to ``Solver(specs_to_add=..., specs_to_remove=...)``
+        with a custom channel containing the data specified in `packages`.
+        """
+        with tempfile.TemporaryDirectory(prefix='conda-test-repo-') as tmpdir_path:
+            tmpdir = pathlib.Path(tmpdir_path)
+            channels = tuple(
+                Channel(f'file://{tmpdir / subdir}')
+                for subdir in context.subdirs
+            )
+            self._write_channel_packages(tmpdir, context.subdirs, list(packages))
+            yield self.solver_class(
+                prefix='dummy - does not exist',
+                subdirs=context.subdirs,
+                channels=channels,
+                specs_to_add=add,
+                specs_to_remove=remove,
+            )
+
+    def install(self, *specs, packages=None):
+        """Try a install transaction in channel with a set of specified packages."""
+        if not packages:
+            packages = self.index(1)
+        with self.solver(add=specs, packages=packages) as solver:
+            return solver.solve_final_state()
+
+    def index(self, num):
+        """Get the index data of the ``helpers.get_index_r_*`` helpers."""
+        get_index = getattr(helpers, f'get_index_r_{num}')
+        index, _ = get_index(context.subdir)
+        return index.values()
+
+    def package_string_set(self, packages):
+        """Transforms package container in package string set."""
+        return {
+            f'{record.name}-{record.version}-{record.build}'
+            for record in packages
+        }
+
+    def assert_installed(self, specs, expecting):
+        """Helper to assert that a transaction result contains the packages
+        specified by the set of specification string."""
+        installed = self.install(*specs)
+        assert installed, f'no installed specs ({installed})'
+        assert self.package_string_set(installed) == set(expecting) | {
+            # XXX: injected
+            'distribute-0.6.36-py27_1',
+            'pip-1.3.1-py27_1',
+        }
+
+    def assert_same_installed(self, specs1, specs2):
+        assert self.package_string_set(
+            self.install(*specs1),
+        ) == self.package_string_set(
+            self.install(*specs2),
+        )
+
+    def assert_record_in(self, record_str, records):
+        """Helper to assert that a record list contains a record matching the
+        provided record string."""
+        assert record_str in [
+            f'{record.name}-{record.version}-{record.build}' for record in records
```

Comment:
I think this should try to use the `_pkey` attribute since that includes channel name, subdir and buildnumber in addition to the fields listed here already.

Reply:
I specifically wanted to remove the channel name as it is the temporary directory.

Reply:
What's the expected behaviour if a solution contains packages from more than one channel?

Reply:
Do we have many tests that need to make sure certain packages were installed from a specific channel? I was thinking we could have a custom helper for that, because most tests don't actually use multiple channels.

Reply:
Yeah, fwiw, it's pretty common to mix the defaults channel with other channels, so I think we shouldn't optimize for single-channel-installs but try to figure out how to always take channels into account.